plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/on-sale.tsx

```diff
@@ -59,8 +63,26 @@ const innerBlocks: InnerBlockTemplate[] = [
 	INNER_BLOCKS_PRODUCT_TEMPLATE,
 ];
 
+/**
+ * This is Async operation example to handle preview state.
+ * I am just using setTimeout to simulate async operation, but it could be
+ * any async operation like fetching data from server or doing a polling in every 5 seconds.
+ */
+const handlePreviewState = ( { setPreviewState }: HandlePreviewStateArgs ) => {
+	setTimeout( () => {
```

Comment:
Assuming long polling like this is a use-case you want to support it might make sense to also support clean up as a return value.

In this function you'd return a function that does the clean up, which you can then return from the useEffect cleanup.

e.g. at the end of this function:

```
return () => {
  clearTimeout(timeoutId);
 }
 ```
 
 then in the component:
 
 ```
 	useEffect( () => {
		const cleanup = handlePreviewState?.( {
			setPreviewState,
		} );
        if (cleanup) { return cleanup; }
	}, [] );
```

Reply:
Good point! 

> Assuming long polling like this is a use-case you want to support it might make sense to also support clean up as a return value.
> 
> In this function you'd return a function that does the clean up, which you can then return from the useEffect cleanup.
> 
> e.g. at the end of this function:
> 
> ```
> return () => {
>   clearTimeout(timeoutId);
>  }
> ```
> 
> then in the component:
> 
> ```
> 	useEffect( () => {
>    	const cleanup = handlePreviewState?.( {
>    		setPreviewState,
>    	} );
>        if (cleanup) { return cleanup; }
>    }, [] );
> ```

Great point! I will add a cleanup functionality.